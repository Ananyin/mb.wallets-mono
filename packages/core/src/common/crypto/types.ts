import { CommonKey } from "common/types/key";

export type CryptoHelper = {
  buildSignSuite: (keyOptions: BuildSignSignatureOptions) => Object

  hash: (data: string) => string

  sign: (data: string, key: string) => string

  getRandomBytes: (size: number) => Promise<Uint8Array>

  verify: (signature: string, data: string, key: string) => boolean

  normalizePassword: (password: string) => Uint8Array

  encrypt: (body: string, password: string) => Promise<string>

  decrypt: (chiper: string, password: string)=> Promise<string>

  makeDerivationPath: (index?: number, change?: number, account?: number, bc?: string) => string

  makeId: (key: string, payload?: string, expand?: boolean) => string

  /**
   * @bug seed param implicitly requires Buffer not Uint8Array!
   */
  getKey: (seed: Uint8Array, derivationPath?: string) => CommonKey & { dp: string }
}

export type BuildSignSignatureOptions = {
  publicKey: string
  privateKey: string
  id: string
  controller: string
}