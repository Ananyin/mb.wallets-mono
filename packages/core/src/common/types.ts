import { TContext } from "@affinidi/vc-common"
import { CryptoHelper } from "metabelarusid-common"
import { KeyChainWrapper } from "keys/types"
import { CommonCredentail, CommonCredentailSubject, CommonSubjectType, CommonType, CommonUnsignedCredential } from "./types/credential"
import { CryptoKey } from "metabelarusid-common"

export type BuildCommonContextMethod = (options: {
  keys: KeyChainWrapper
  crypto: CryptoHelper
}) => Promise<CommonContext>

export type CommonContext = {
  keys: KeyChainWrapper
  crypto: CryptoHelper
  buildCredential: CommonBuildCredentialMethod
  signCredential: CommonSignCredentialMethod
  verifyCredential: CommonVerfiyCredentailMethod
}

export type CommonBuildCredentialMethod = <
  SubjectType extends CommonSubjectType = CommonSubjectType,
  Subject extends CommonCredentailSubject<SubjectType> = CommonCredentailSubject<SubjectType>
  >(options: CommonBuildCredentailOptions<SubjectType>) =>
  Promise<CommonUnsignedCredential<Subject>>

export type CommonContextType = TContext

export type CommonBuildCredentailOptions<
  SubjectType extends CommonSubjectType = CommonSubjectType,
  Subject extends CommonCredentailSubject<SubjectType> = CommonCredentailSubject<SubjectType>
  > = {
    id: string,
    type: CommonType
    holder: string,
    subject: Subject,
    issueanceDate?: string
    context: TContext
  }

export type CommonSignCredentialMethod =
  <
    Subject extends CommonCredentailSubject = CommonCredentailSubject
    >(
    unsingedCredential: CommonUnsignedCredential<Subject>,
    issuer: string,
    key: CryptoKey,
    options?: CommonSignCredentialOptions
  ) =>
    Promise<CommonCredentail<Subject>>

export type CommonSignCredentialOptions = {
  controllerRole?: ControllerRole
  buildProofPurposeOptions?: () => Promise<Object>
}

export type CommonVerfiyCredentailMethod = (credential: CommonCredentail) => Promise<boolean>

export type ControllerRole =
  typeof COMMON_CONTROLLER_ROLE_ISSUER
  | typeof COMMON_CONTROLLER_ROLE_HOLDER

export const COMMON_CONTROLLER_ROLE_ISSUER = 'issuer'
export const COMMON_CONTROLLER_ROLE_HOLDER = 'holder'