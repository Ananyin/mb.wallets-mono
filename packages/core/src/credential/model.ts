
import { _keyChainHelper } from "common/helper/key";
import { CommonContext } from "common/types";
import { CreateCredentialMethod, ERROR_NO_HOLDER, ERROR_NO_ISSUER, SignCredentialMethod } from "./types";


export const buildCreateCrednetialMethod =
  (context: CommonContext): CreateCredentialMethod =>
    async (type, subject, holder = undefined) => {
      if (holder === undefined) {
        throw new Error(ERROR_NO_HOLDER)
      }

      return context.buildCredential({
        id: 'did:peer:xxxxx', // @TODO generate ID from context
        type,
        subject,
        holder
      })
    }

export const buildSignCredentialMethod =
  (context: CommonContext): SignCredentialMethod =>
    async (credential, issuer?, options = {}) => {
      if (!issuer) {
        throw new Error(ERROR_NO_ISSUER)
      }

      const key = _keyChainHelper.parseSigningKeyOptions(context, options)
      return context.signCredential(
        credential,
        issuer,
        await _keyChainHelper.keyToCommonKey(
          key,
          options.password || context.keyChain.getDefaultPassword(),
          { rotation: options.rotation }
        ),
        { controllerRole: options.controllerRole }
      )
    }