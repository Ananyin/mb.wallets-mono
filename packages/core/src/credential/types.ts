import { CommonContextType } from "./context/types"
import { CommonCredentail, CommonCredentailSubject, CommonSubjectType, CommonType, CommonUnsignedCredential } from "./context/types/credential"


export type Credential<Subject extends CredentialSubject = CredentialSubject>
  = CommonCredentail<Subject> & {}

export type Identity<Subject extends IdentitySubect = IdentitySubect>
  = {} & Credential<Subject>

export type IdentitySubect<Type extends CredentialSubjectType = CredentialSubjectType> = CredentialSubject<Type>

export type CredentialSubjectProperty<Type extends CredentialSubject = CredentialSubject>
  = Type

export type CredentialSubject<SubjectType extends CredentialSubjectType = CredentialSubjectType> =
  CommonCredentailSubject<SubjectType> & {}

export type CredentialSubjectType = CommonSubjectType & {}

export type CredentialContextType = CommonContextType & {}

export type UnsignedCredentail<
  Subject extends CredentialSubject = CredentialSubject
  > = CommonUnsignedCredential<Subject> & {}

export type CredentialType = CommonType