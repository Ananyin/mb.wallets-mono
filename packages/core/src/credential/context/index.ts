export * from './types'

export * from './types/index'