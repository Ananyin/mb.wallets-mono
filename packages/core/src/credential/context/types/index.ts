
export * from './credential'
// export * from './presentation'