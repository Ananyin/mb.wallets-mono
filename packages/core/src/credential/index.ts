
export * from './types'
export * from './context/index'
export * from './context'

