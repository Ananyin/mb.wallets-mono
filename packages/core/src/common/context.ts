
import { buildVCV1, buildVCV1Skeleton, buildVCV1Unsigned } from "@affinidi/vc-common";

import { BuildCommonContextMethod, CommonBuildCredentailOptions, CommonSignCredentialOptions, COMMON_CONTROLLER_ROLE_HOLDER } from "./types";
import { CommonCredentail, CommonCredentailSubject, CommonSubjectType, CommonUnsignedCredential } from "./types/credential";
import { CommonKey } from "./types/key";

export const buildCommonContext: BuildCommonContextMethod = async ({
  keyChain,
  cryptoContext
}) => {
  const documentLoader = async (url: string): Promise<any> => {
    if (url.startsWith('did:')) {
      // @TODO Fix lookup by did for metabelarus purpose
      throw new SyntaxError('Can\'t look up did based urls fore documents')
      return {
        contextUrl: null,
        document: {},
        documentUrl: url,
      }
    }

    const jsonld = require('jsonld')

    return jsonld.documentLoaders.node()(url)
  }

  return {
    keyChain,

    cryptoContext,

    buildCredential: async <
      T extends CommonSubjectType = CommonSubjectType,
      S extends CommonCredentailSubject<T> = CommonCredentailSubject<T>
    >(options: CommonBuildCredentailOptions<T>) => {
      const skeleton = buildVCV1Skeleton({
        id: options.id,
        type: options.type,
        credentialSubject: options.subject,
        holder: {
          id: options.holder
        },
        context: options.context
      })

      return buildVCV1Unsigned({
        skeleton,
        issuanceDate: options.issueanceDate || (new Date).toISOString()
      }) as CommonUnsignedCredential<S>
    },

    signCredential: async <
      S extends CommonCredentailSubject = CommonCredentailSubject
    >(
      unsingedCredential: CommonUnsignedCredential<S>,
      issuer: string,
      key: CommonKey,
      options?: CommonSignCredentialOptions
    ) => {
      try {
        return await buildVCV1({
          unsigned: unsingedCredential,
          issuer: {
            did: issuer,
            keyId: key.id,
            privateKey: key.pk,
            publicKey: key.pubKey
          },
          getSignSuite: cryptoContext.buildSignSuite,
          documentLoader,
          getProofPurposeOptions: options?.buildProofPurposeOptions || (async () => ({
            controller: {
              id: !options?.controllerRole || options.controllerRole
                === COMMON_CONTROLLER_ROLE_HOLDER ? unsingedCredential.holder.id : issuer
            }
          }))
        }) as CommonCredentail<S>
      } catch (e) {
        console.log(e.details)

        throw e
      }
    }
  }
}