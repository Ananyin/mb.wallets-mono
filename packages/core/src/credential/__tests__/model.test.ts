import { buildCommonContext } from "common/context"
import { CommonContext } from "common/types"
import { buildCreateCrednetialMethod, buildSignCredentialMethod } from "credential/model"
import { CredentialSubjectType, UnsignedCredentail } from "credential/types"
import { nodeCryptoContext } from "crypto/context/node"
import { buildKeyChain } from "keys/model"

import util from 'util'
util.inspect.defaultOptions.depth = 8

const testContext: {
  commonContext?: CommonContext
  unsignedCredential?: UnsignedCredentail
} = {}

beforeAll(async () => {
  testContext.commonContext = await buildCommonContext({
    keyChain: await buildKeyChain('11111111'),
    cryptoContext: nodeCryptoContext
  })
})

describe('Credential Model', () => {
  it('Create Credentail', async () => {
    if (!testContext.commonContext) {
      throw 'Setup didn\'t provide CommonContext'
    }
    const createCredential = buildCreateCrednetialMethod(testContext.commonContext)
    const unsingnedCredentail = await createCredential<
      CredentialSubjectType & { worker: string }
    >(
      ['VerifiableCredential', 'TestCredential'],
      {
        data: {
          '@type': 'TestCredentialSubjectDataType',
          worker: 'Valentin Michalych'
        }
      },
      'did:peer:yyyy',
      {
        '@version': 1.1,
        meta: 'https://meta-id.meta-belarus.org/vc-schema#',
        data: {
          '@id': 'meta:data',
          '@type': '@id',
          '@context': {
            worker: { '@id': 'meta:worker', '@type': 'xsd:string' }
          }
        }
      }
    )

    testContext.unsignedCredential = unsingnedCredentail
    expect(unsingnedCredentail).toMatchSnapshot({
      issuanceDate: expect.any(String)
    })
  })

  it('Sing Credential', async () => {
    if (!testContext.commonContext) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!testContext.unsignedCredential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }
    const signCredential = buildSignCredentialMethod(testContext.commonContext)
    const credentail = await signCredential(testContext.unsignedCredential, 'did:peer:zzzzz')

    expect(credentail).toMatchSnapshot({
      issuanceDate: expect.any(String),
      proof: {
        created: expect.any(String),
        jws: expect.any(String),
        verificationMethod: expect.any(String)
      }
    })
  })
})