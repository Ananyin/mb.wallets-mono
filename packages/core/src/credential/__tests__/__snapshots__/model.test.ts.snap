// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credential Model Create Credentail 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "worker": Object {
            "@id": "meta:worker",
            "@type": "xsd:string",
          },
        },
        "@id": "meta:data",
        "@type": "@id",
      },
      "meta": "https://meta-id.meta-belarus.org/vc-schema#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "TestCredentialSubjectDataType",
      "worker": "Valentin Michalych",
    },
  },
  "holder": Object {
    "id": "did:peer:yyyy",
  },
  "id": "did:peer:xxxxx",
  "issuanceDate": Any<String>,
  "type": Array [
    "VerifiableCredential",
    "TestCredential",
  ],
}
`;

exports[`Credential Model Sing Credential 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "worker": Object {
            "@id": "meta:worker",
            "@type": "xsd:string",
          },
        },
        "@id": "meta:data",
        "@type": "@id",
      },
      "meta": "https://meta-id.meta-belarus.org/vc-schema#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "TestCredentialSubjectDataType",
      "worker": "Valentin Michalych",
    },
  },
  "holder": Object {
    "id": "did:peer:yyyy",
  },
  "id": "did:peer:xxxxx",
  "issuanceDate": Any<String>,
  "issuer": "did:peer:zzzzz",
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "assertionMethod",
    "type": "EcdsaSecp256k1Signature2019",
  },
  "type": Array [
    "VerifiableCredential",
    "TestCredential",
  ],
}
`;
